src/file.rs

```diff
@@ -0,0 +1,59 @@
+use anyhow::{Context, Result};
+use std::fs;
+use std::io;
+use std::path::Path;
+
+pub(crate) fn hardlink(original: &Path, link: &Path) -> io::Result<()> {
+    let _ = fs::remove_file(link);
+    fs::hard_link(original, link)
+}
+
+pub(crate) fn hard_or_symlink_file(original: &Path, link: &Path) -> Result<()> {
+    if hardlink_file(original, link).is_err() {
+        symlink_file(original, link)?;
+    }
+    Ok(())
+}
+
+pub fn hardlink_file(original: &Path, link: &Path) -> Result<()> {
+    hardlink(original, link).with_context(|| {
+        format!(
+            "Could not create link: {}->{}",
+            original.display(),
+            link.display()
+        )
+    })
+}
+
+#[cfg(unix)]
+fn symlink_file(original: &Path, link: &Path) -> Result<()> {
+    std::os::unix::fs::symlink(original, link).with_context(|| {
+        format!(
+            "Could not create link: {}->{}",
+            original.display(),
+            link.display()
+        )
+    })
+}
+
+pub fn is_file<P: AsRef<Path>>(path: P) -> bool {
+    fs::metadata(path).ok().as_ref().map(fs::Metadata::is_file) == Some(true)
+}
```

Comment:
Seems like the only difference is Path::is_file() uses `map` instead of `ok` - are there any real consequences to this?

I tested a small snippet out locally and both yield the same result. I chose to use the same code as rustup in case there was anything i might not have been considering, but from what I see it should be ok to refactor this to use Path::is_file() instead

Reply:
Seems like the only difference is whether we want to handle it as an `Option` (in ok) or just `map` and leave Errs untouched. Will refactor this^